### PR TITLE
Issue #1404: implemented tests for ClusterApiService

### PR DESCRIPTION
**Pull request by youkofan**
First variant of tests for the first method (with @PostFilter) in the ClusterApiService
Issue - https://github.com/epam/cloud-pipeline/issues/1404
Signed-off-by: Iurii_Kofanov <youkofan@gmail.com>

### CODE REVIEW

File: api/src/test/java/com/epam/pipeline/app/AclTestConfiguration.java [hunk 1]
```diff
@@ -144,6 +146,12 @@
     @MockBean
     protected CloudRegionManager mockCloudRegionManager;
 
+    @MockBean
+    protected NodeDiskManager mockNodeDiskManager;
+
+    @MockBean
+    UsageMonitoringManager mockUsageMonitoringManager;
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/app/AclTestConfiguration.java]: Could you please add access modifier?

File: api/src/test/java/com/epam/pipeline/test/acl/AclTestBeans.java [hunk 2]
```diff
@@ -322,9 +438,9 @@ protected DataStorageManager mockDataStorageManager() {
     }
 
     @Bean
-    protected FolderCrudManager mockCrudManager() {
+    protected FolderCrudManager mockFolderCrudManager() {
         return Mockito.mock(FolderCrudManager.class);
-    }
+    } /////
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/test/acl/AclTestBeans.java]: Cleanup please

File: api/src/test/java/com/epam/pipeline/test/acl/AclTestConfiguration.java [hunk 3]
```diff
@@ -39,7 +41,11 @@
 @Target(ElementType.TYPE)
 @Retention(RetentionPolicy.RUNTIME)
 @ContextConfiguration(classes = {AclTestBeans.class, AclSecurityConfiguration.class})
-@ComponentScan(basePackages = {"com.epam.pipeline.security.acl", "com.epam.pipeline.manager.security"})
+@Import({ContextualPreferenceHandler.class})
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/test/acl/AclTestConfiguration.java]: Why we need such import?

File: api/src/test/java/com/epam/pipeline/test/acl/AclTestConfiguration.java [hunk 4]
```diff
@@ -39,7 +41,11 @@
...
+@ComponentScan(basePackages = {"com.epam.pipeline.security.acl",
+        "com.epam.pipeline.security.run",
+        "com.epam.pipeline.manager.security",
+        "com.epam.pipeline.manager.user"})
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/test/acl/AclTestConfiguration.java]: Why we need `UserManager` real bean?

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+    private UsageMonitoringManager mockUsageMonitoringManager;
+
+    @Autowired
+    private NodeDiskManager mockNodeDiskManager;
+
+    @Autowired
+    private InstanceOfferManager mockInstanceOfferManager;
+
+    private final FilterPodsRequest filterPodsRequest = NodeCreatorUtils.getDefaultFilterPodsRequest();
+
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Let's remove empty lines between fields.

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+        statsList = Collections.singletonList(monitoringStats);
+
+        instanceTypes = Collections.singletonList(instanceType);
+
+        nodeDisks = Collections.singletonList(nodeDisk);
+
+        pipelineRun.setId(1L);
+        pipelineRun.setOwner(SIMPLE_USER);
+
+        final PipelineRun pipelineRunWithoutPermission = NodeCreatorUtils.getPipelineRun();
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Let's create a new build method in creators package. This  method may accept `id`, `name`, `owner` arguments.

File: api/src/test/java/com/epam/pipeline/test/creator/cluster/NodeCreatorUtils.java [hunk 7]
```diff
@@ -86,4 +99,20 @@ public static AllowedInstanceAndPriceTypes getDefaultAllowedInstanceAndPriceType
     public static NodeDisk getDefaultNodeDisk() {
         return new NodeDisk(ID, TEST_STRING, LDT);
     }
+
+    public static PipelineRun getPipelineRun() {
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/test/creator/cluster/NodeCreatorUtils.java]: Let's not confuse packages: `PipelineRun` is not in cluster package. We need to create a new package/class for `PipelineRun` object creating methods.

Comment (ekazachkova) [reply api/src/test/java/com/epam/pipeline/test/creator/cluster/NodeCreatorUtils.java]: The same for creators below

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+
+    private List<MonitoringStats> statsList;
+
+    private List<NodeInstance> singleNodeInstance;
+
+    private List<NodeInstance> twoNodeInstances;
+
+    @Before
+    public void setUp() {
+        statsList = Collections.singletonList(monitoringStats);
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Let's  inline test class field initiliazations and get rid of `setUp` method.

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,546 @@
...
+        twoNodeInstances.add(nodeInstance);
+        twoNodeInstances.add(nodeInstanceWithoutPermission);
+    }
+
+    @Test
+    @WithMockUser(roles = ADMIN_ROLE)
+    public void shouldReturnListWithNodeInstancesForAdmin() {
+        doReturn(singleNodeInstance).when(mockNodesManager).getNodes();
+
+        List<NodeInstance> nodes = clusterApiService.getNodes();
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Do not forget final please

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,546 @@
...
+        doReturn(SIMPLE_USER).when(mockAuthManager).getAuthorizedUser();
+        initAclEntity(nodeInstance,
+                Collections.singletonList(new UserPermission(SIMPLE_USER_ROLE, AclPermission.READ.getMask())));
+        doReturn(contextualPreference).when(mockContextualPreferenceManager).search(
+                Collections.singletonList(SystemPreferences.RUN_VISIBILITY_POLICY.getKey()));
+        doReturn(singleNodeInstance).when(mockNodesManager).getNodes();
+
+        List<NodeInstance> nodes = clusterApiService.getNodes();
+
+        assertThat(nodes).hasSize(1);
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Could we use `contains` method here?

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,546 @@
...
+
+    private final FilterNodesVO filterNodesVO = NodeCreatorUtils.getDefaultFilterNodesVO();
+
+    private final NodeDisk nodeDisk = NodeCreatorUtils.getDefaultNodeDisk();
+
+    private final MonitoringStats monitoringStats = NodeCreatorUtils.getMonitoringStats();
+
+    private final InstanceType instanceType = NodeCreatorUtils.getDefaultInstanceType();
+
+    private InputStream inputStream;
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Looks like this field was not initialized

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java (hunk 5 above)
Comment (ekazachkova) [reply api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Also, cold you please place fields autowired fields declaration?

**Review by ekazachkova — CHANGES_REQUESTED** [review]

File: api/src/test/java/com/epam/pipeline/test/creator/cluster/NodeCreatorUtils.java [hunk 12]
```diff
@@ -51,6 +53,24 @@ public static NodeInstance getDefaultNodeInstance() {
         return new NodeInstance();
     }
 
+    public static NodeInstance getNodeInstanceWithPermission() {
+        NodeInstance nodeInstance = new NodeInstance();
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/test/creator/cluster/NodeCreatorUtils.java]: Could you please add final to fields here and in other places?

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,508 @@
...
+    @Autowired
+    private UsageMonitoringManager mockUsageMonitoringManager;
+
+    @Autowired
+    private NodeDiskManager mockNodeDiskManager;
+
+    @Autowired
+    private InstanceOfferManager mockInstanceOfferManager;
+
+    private List<NodeInstance> initSingleNodeInstanceList() {
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: The `private` methods shell be after all `public` methods

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,508 @@
...
+    public void shouldReturnStatsForNodeForAdmin() {
+        final MonitoringStats monitoringStats = ContextualPreferenceCreatorUtils.getMonitoringStats();
+        final List<MonitoringStats> statsList = Collections.singletonList(monitoringStats);
+        doReturn(statsList).when(mockUsageMonitoringManager)
+                .getStatsForNode(nodeInstance.getName(), LocalDateTime.MIN, LocalDateTime.MAX);
+
+        final List<MonitoringStats> resultStatsList =
+                clusterApiService.getStatsForNode(nodeInstance.getName(), LocalDateTime.MIN, LocalDateTime.MAX);
+
+        assertThat(resultStatsList.size()).isEqualTo(1);
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Could we use here
```
assertThat(resultStatsList).hasSize(1).contains(monitoringStats);
```
And in other such places too.

File: api/src/test/java/com/epam/pipeline/test/acl/AclTestConfiguration.java [hunk 15]
```diff
@@ -39,7 +39,10 @@
 @Target(ElementType.TYPE)
 @Retention(RetentionPolicy.RUNTIME)
 @ContextConfiguration(classes = {AclTestBeans.class, AclSecurityConfiguration.class})
-@ComponentScan(basePackages = {"com.epam.pipeline.security.acl", "com.epam.pipeline.manager.security"})
+@ComponentScan(basePackages = {"com.epam.pipeline.security.acl",
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/test/acl/AclTestConfiguration.java]: It seems that we can just use `basePackages = {"com.epam.pipeline.security.acl", "com.epam.pipeline.manager.security"}`

File: api/src/test/java/com/epam/pipeline/test/creator/cluster/NodeCreatorUtils.java [hunk 16]
```diff
@@ -23,6 +23,7 @@
 import com.epam.pipeline.entity.cluster.MasterNode;
 import com.epam.pipeline.entity.cluster.NodeDisk;
 import com.epam.pipeline.entity.cluster.NodeInstance;
+import com.epam.pipeline.test.creator.pipeline.PipelineCreatorUtils;
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/test/creator/cluster/NodeCreatorUtils.java:29]: Could you please rename this class to `ClusterCreatorUtils`? This class should consist of creators for entities from `cluster` package.

File: api/src/test/java/com/epam/pipeline/test/creator/contextual/ContextualPreferenceCreatorUtils.java [hunk 17]
```diff
@@ -55,4 +56,8 @@ public static ContextualPreferenceVO getContextualPreferenceVO() {
                 TEST_STRING, TEST_STRING, PREFERENCE_TYPE, getCPExternalResource()
         );
     }
+
+    public static MonitoringStats getMonitoringStats() {
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/test/creator/contextual/ContextualPreferenceCreatorUtils.java]: Could you move this method to `ClusterCreatorUtils`?

File: api/src/test/java/com/epam/pipeline/test/creator/pipeline/PipelineCreatorUtils.java [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+
+import com.epam.pipeline.entity.pipeline.PipelineRun;
+
+import static com.epam.pipeline.test.creator.CommonCreatorConstants.ID;
+import static com.epam.pipeline.test.creator.CommonCreatorConstants.ID_2;
+import static com.epam.pipeline.test.creator.CommonCreatorConstants.TEST_STRING;
+
+public final class PipelineCreatorUtils {
+
+    public static PipelineRun getPipelineRunWithPermission() {
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/test/creator/pipeline/PipelineCreatorUtils.java]: It is absolutely not clear what is meant `permission`

File: api/src/test/java/com/epam/pipeline/test/creator/pipeline/PipelineCreatorUtils.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+package com.epam.pipeline.test.creator.pipeline;
+
+import com.epam.pipeline.entity.pipeline.PipelineRun;
+
+import static com.epam.pipeline.test.creator.CommonCreatorConstants.ID;
+import static com.epam.pipeline.test.creator.CommonCreatorConstants.ID_2;
+import static com.epam.pipeline.test.creator.CommonCreatorConstants.TEST_STRING;
+
+public final class PipelineCreatorUtils {
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/test/creator/pipeline/PipelineCreatorUtils.java:23]: Let's avoid some test case specific naming in such classes. For example, for current situation we need only one method  for this class - `getPipelineRun` that will return `PipelineRun` object and accepts all required argument: id, name, owner... You may take the `ObjectCreatorUtils` class as a basis.

File: api/src/test/java/com/epam/pipeline/test/creator/cluster/NodeCreatorUtils.java [hunk 20]
```diff
@@ -51,6 +53,24 @@ public static NodeInstance getDefaultNodeInstance() {
...
+    public static NodeInstance getNodeInstanceWithPermission() {
+        NodeInstance nodeInstance = new NodeInstance();
+        nodeInstance.setId(ID);
+        nodeInstance.setOwner("OWNER");
+        nodeInstance.setPipelineRun(PipelineCreatorUtils.getPipelineRunWithPermission());
+        nodeInstance.setName(TEST_STRING);
+        return nodeInstance;
+    }
+
+    public static NodeInstance getNodeInstanceWithoutPermission() {
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/test/creator/cluster/NodeCreatorUtils.java]: see `PipelineCreatorUtils` comment

File: api/src/test/java/com/epam/pipeline/test/creator/pipeline/PipelineCreatorUtils.java [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+
+    public static PipelineRun getPipelineRunWithPermission() {
+        PipelineRun pipelineRun = new PipelineRun();
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/test/creator/pipeline/PipelineCreatorUtils.java]: Do not forget to make variables `final`

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 22]
```diff
@@ -0,0 +1,508 @@
...
+        doReturn(nodeInstance).when(mockNodesManager).getNode(nodeInstance.getName());
+        doReturn(pipelineRun).when(mockPipelineRunManager).loadPipelineRun(eq(pipelineRun.getId()));
+
+        clusterApiService.terminateNode(nodeInstance.getName());
+    }
+
+    @Test
+    @WithMockUser(roles = ADMIN_ROLE)
+    public void shouldReturnStatsForNodeForAdmin() {
+        final MonitoringStats monitoringStats = ContextualPreferenceCreatorUtils.getMonitoringStats();
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: It seems this initialization is redundant.

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 23]
```diff
@@ -0,0 +1,508 @@
...
+
+        assertThat(nodes).isEmpty();
+    }
+
+    @Test
+    @WithMockUser(roles = ADMIN_ROLE)
+    public void shouldReturnNodeInstanceForAdmin() {
+        doReturn(nodeInstance).when(mockNodesManager).getNode(nodeInstance.getName());
+
+       final NodeInstance node = clusterApiService.getNode(nodeInstance.getName());
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Format code please

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 24]
```diff
@@ -0,0 +1,508 @@
...
+        twoNodeInstances.add(nodeInstanceWithoutPermission);
+        return twoNodeInstances;
+    }
+
+    @Test
+    @WithMockUser(roles = ADMIN_ROLE)
+    public void shouldReturnListWithNodeInstancesForAdmin() {
+        doReturn(initSingleNodeInstanceList()).when(mockNodesManager).getNodes();
+
+        final List<NodeInstance> nodes = clusterApiService.getNodes();
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: It seems for such simple and short cases we can merge these two lines: 
```
assertThat(clusterApiService.getNodes()).hasSize(1).contains(nodeInstance);
```

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 25]
```diff
@@ -0,0 +1,508 @@
...
+
+        assertThat(nodes).hasSize(1).contains(nodeInstance);
+    }
+
+    @Test
+    @WithMockUser(username = SIMPLE_USER)
+    public void shouldReturnListWithNodeInstancesWhenPermissionIsGranted() {
+        doReturn(SIMPLE_USER).when(mockAuthManager).getAuthorizedUser();
+        initAclEntity(nodeInstance,
+                Collections.singletonList(new UserPermission(SIMPLE_USER_ROLE, AclPermission.READ.getMask())));
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Maybe we shell consider creating a common for `*ApiServiceTest` classes method to build a single user permissions?

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 26]
```diff
@@ -0,0 +1,508 @@
...
+        doReturn(contextualPreference).when(mockContextualPreferenceManager).search(
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Looks like this initialization is redundant

Comment (ekazachkova) [reply api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Could you check it for other methods too?

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java (hunk 24 above)
Comment (ekazachkova) [reply api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Could you check it for other methods too?

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 27]
```diff
@@ -0,0 +1,508 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static com.epam.pipeline.test.creator.CommonCreatorConstants.TEST_STRING;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.doReturn;
+
+public class ClusterApiServiceTest extends AbstractAclTest {
```
Comment (ekazachkova) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java:61]: Maybe we shell add `@AclMask` tests? (for merging entity permissions with parent permissions)

P.S.: see `AclAspect`

**Review by ekazachkova — CHANGES_REQUESTED** [review]

File: api/src/test/java/com/epam/pipeline/test/acl/AclTestBeans.java [hunk 28]
```diff
@@ -17,47 +17,79 @@
 package com.epam.pipeline.test.acl;
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/test/acl/AclTestBeans.java]: Could you please check if all of the changes in this class are relevant to the pull request? It seems that mocks like `PipelineDao` have nothing to do with acl tests.

File: api/src/main/java/com/epam/pipeline/entity/cluster/NodeInstance.java [hunk 29]
```diff
@@ -45,7 +45,6 @@
 public class NodeInstance extends AbstractSecuredEntity {
 
     private UUID uid;
-    private String name;
```
Comment (tcibinan) [inline api/src/main/java/com/epam/pipeline/entity/cluster/NodeInstance.java]: I'm afraid that this change can be harmful if some api method has `NodeInstance` class in request body. For example as part of `PipelineRun` class. Let's revert the change until we have implemented tests for all the controllers.

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 30]
```diff
@@ -0,0 +1,451 @@
...
+import java.util.List;
+
+import static com.epam.pipeline.util.CustomAssertions.assertThrows;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.doReturn;
+
+public class ClusterApiServiceTest extends AbstractAclTest {
+
+    private final String TEST_STRING = "TEST";
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Could you please replace the field usages with `CommonCreatorConstants.TEST_STRING`.

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 31]
```diff
@@ -0,0 +1,451 @@
...
+    private final FilterPodsRequest filterPodsRequest = NodeCreatorUtils.getDefaultFilterPodsRequest();
+    private final NodeInstance nodeInstance = NodeCreatorUtils.getNodeInstance(1L, SIMPLE_USER);
+    private final NodeInstance nodeInstanceWithoutPermission = NodeCreatorUtils.getNodeInstance(2L, TEST_STRING);
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: I suppose we can call the second one just `anotherNodeInstance`.

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 32]
```diff
@@ -0,0 +1,451 @@
...
+        assertThat(clusterApiService.getNodes()).hasSize(1).contains(nodeInstance);
+    }
+
+    @Test
+    @WithMockUser
+    public void shouldReturnEmptyNodeInstanceListWhenPermissionIsNotGranted() {
+        initAclEntity(nodeInstanceWithoutPermission);
+        doReturn(initSingleNodeInstanceList()).when(mockNodesManager).getNodes();
+
+        assertThat(clusterApiService.getNodes()).isEmpty();
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Method `initSingleNodeInstanceList()` returns list which doesn't contain `nodeInstanceWithoutPermission` so this test is not testing what it should. Probably we can even get rid of methods like `initSingleNodeInstanceList` and always initialiaze lists like the following
```java
Collections.singletonList(node);
// or
Arrays.asList(node1, node2);
```

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 33]
```diff
@@ -0,0 +1,451 @@
...
+
+        final NodeInstance returnedNodeInstance = clusterApiService.getNode(nodeInstance.getName());
+
+        assertThat(returnedNodeInstance).isEqualTo(nodeInstance);
+        assertThat(returnedNodeInstance.getMask()).isEqualTo(AclPermission.READ.getMask());
+    }
+
+    @Test
+    @WithMockUser
+    public void shouldDenyAccessToNode() {
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Let's specify explicitly why the access is denied like `shouldDenyAccessToNodeWhenPermissionIsNotGranted`.

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 34]
```diff
@@ -0,0 +1,451 @@
...
+    @Test
+    @WithMockUser
+    public void shouldReturnStatsWhenPermissionIsGranted() {
+        doReturn(SIMPLE_USER).when(mockAuthManager).getAuthorizedUser();
+        initAclEntity(nodeInstance, AclPermission.READ);
+        doReturn(statsList).when(mockUsageMonitoringManager).getStatsForNode(
+                nodeInstance.getName(), LocalDateTime.MIN, LocalDateTime.MAX);
+        doReturn(nodeInstance).when(mockNodesManager).getNode(nodeInstance.getName(), filterPodsRequest);
+        doReturn(nodeInstance).when(mockNodesManager).getNode(nodeInstance.getName());
+        doReturn(pipelineRun).when(mockPipelineRunManager).loadPipelineRun(eq(pipelineRun.getId()));
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: I wonder if we can extract all this duplicated mocking calls into some methods like the following:
```java
mockUser();
mockStats();
mockNodeInstance();
mockPipelineRun();
```

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java (hunk 30 above)
Comment (youkofan) [reply api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: I already have four static imports in this class, and, if i'm not mistaken, PMD avoid amount of static imports more than four.

File: api/src/test/java/com/epam/pipeline/test/acl/AbstractAclTest.java [hunk 35]
```diff
@@ -131,4 +132,10 @@ public Sid toSid() {
             return new GrantedAuthoritySid(authorityName);
         }
     }
+
+    protected List<Object> mutableListOf(Object... objects) {
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/test/acl/AbstractAclTest.java]: Could you please use generic type to return `List<T>` rather then `List<Object>`. It probably is a good idea because it increases the possibility that this method will be reused.

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 36]
```diff
@@ -0,0 +1,432 @@
...
+
+        assertThat(clusterApiService.getNode(nodeInstance.getName(), filterPodsRequest)).isEqualTo(nodeInstance);
+    }
+
+    @Test
+    @WithMockUser(username = SIMPLE_USER)
+    public void shouldReturnNodeThroughRequestWhenPermissionIsGranted() {
+        final NodeInstance nodeInstance = NodeCreatorUtils.getNodeInstance(1L, TEST_STRING);
+        initAclEntity(nodeInstance, AclPermission.READ);
+        initMocksBehavior(true);
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: Probably this mocking call makes thing even more complicated then before. For example in this case `nodeInstance` used for acl is not the same instance that will be used in `initMocksBehavior`. Also this `true\false` flag is not intuitive. Just imagine that there will be another boolean parameter.

I still think that it will be better to use separate calls for mocks which will just remove a need to write the whole
`doReturn().when().call()` chain. For example in this case we can use something like the following. It does not contain less lines then the original version. But it is not the point, isn't it? The following code is declarative as plain mockito calls but it is also much simpler. And in case f.e. pipeline run mocking mechanism will have to change in the future then we will just have to change a single method `mockRun`.
```java
mockUser(SIMPLE_USER); // Mock any user related calls that are used in the tests.
mockNode(nodeInstance); // Mock all node instance calls that are used in the tests.
mockRun(pipelineRun); // Mock all pipeline run calls that are used in the tests.
```

```java
public void mockNode(final NodeInstance nodeInstance) {
    doReturn(nodeInstance).when(mockNodesManager).getNode(nodeInstance.getName(), filterPodsRequest);
    doReturn(nodeInstance).when(mockNodesManager).getNode(nodeInstance.getName());
}
```

What do you think?

File: api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java [hunk 37]
```diff
@@ -0,0 +1,444 @@
...
+        assertThat(clusterApiService.loadNodeDisks(nodeDisk.getNodeId())).hasSize(1).contains(nodeDisk);
+    }
+
+    @Test
+    @WithMockUser
+    public void shouldDenyAccessToNodeDisksWhenPermissionIsNotGranted() {
+        initAclEntity(nodeInstance);
+        doReturn(nodeDisks).when(mockNodeDiskManager).loadByNodeId(nodeDisk.getNodeId());
+        doReturn(nodeInstance).when(mockNodesManager).getNode(nodeDisk.getNodeId(), filterPodsRequest);
+        doReturn(nodeInstance).when(mockNodesManager).getNode(nodeDisk.getNodeId());
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/acl/cluster/ClusterApiServiceTest.java]: It does look like we can just replace it with the new method `mockNode`. Some similar cases with `terminateNode` calls in different tests too. You can just add `terminateNode` to `mockNode` and everything seems to be just fine.

**Review by tcibinan — APPROVED** [review]
